### ISSUE TITLE
Let SIGHUP reload OpenSSL certificates

### ISSUE THREAD

**Issue by EssGeeEich** [coturn/coturn#725]
Currently coturn only seems to reload the certificates [on startup](https://github.com/coturn/coturn/blob/57180ab60afcaeb13537e69ae8cb8aefd8f3f546/src/apps/relay/mainrelay.c#L2399) or [on SIGUSR2](https://github.com/coturn/coturn/blob/57180ab60afcaeb13537e69ae8cb8aefd8f3f546/src/apps/relay/mainrelay.c#L2524).

I propose this change so you can use `systemctl reload coturn` on certificate renewal hooks without breaking existing connections nor requiring weird trickeries to send SIGUSR2.

Currently SIGHUP seems to be handled [here](https://github.com/coturn/coturn/blob/c0822dab953ffe2a94228e8e92487dd8a90607c5/src/apps/common/ns_turn_utils.c#L337).

Since you can only have one signal handler, I propose to:

1. Export a new function from ns_turn_utils.c that acts more or less like the current [sighup_callback_handler](https://github.com/coturn/coturn/blob/c0822dab953ffe2a94228e8e92487dd8a90607c5/src/apps/common/ns_turn_utils.c#L319).
2. Create a new signal function in [mainrelay.c](https://github.com/coturn/coturn/blob/57180ab60afcaeb13537e69ae8cb8aefd8f3f546/src/apps/relay/mainrelay.c) and call `reload_ssl_certs` and the new function I mentioned in point 1.
3. Move the [signal function call](https://github.com/coturn/coturn/blob/c0822dab953ffe2a94228e8e92487dd8a90607c5/src/apps/common/ns_turn_utils.c#L337) somewhere in `main`, possibly [here](https://github.com/coturn/coturn/blob/57180ab60afcaeb13537e69ae8cb8aefd8f3f546/src/apps/relay/mainrelay.c#L2399) and insert the new signal handler I mentioned in point 2.

Please excuse the absence of a pull request but I don't have a development environment I could use at the moment.

**Comment by wferi** [comment]
Does `systemctl --signal=SIGUSR2 kill coturn` qualify as weird trickery?

**Comment by misi** [comment]
Many Thanks! I ask your patience.. I will check your proposed changed and get back to you soon.

**Comment by EssGeeEich** [comment]
Sorry, I was unaware of --signal, altough it still _feels_ like weird trickery since you're not really killing a service.

In my opinion the issue should stand, since semantically `systemctl reload coturn` should reload certificates (imo).

If the issue gets closed "as-is", may I ask to put `systemctl --signal=SIGUSR2 kill coturn` in the README?
This way the next person knows what to put in a certificate renewal hook without killing the entire service.

**Comment by misi** [comment]
Don't get us wrong we need to reconsider how we could make it..

**Comment by hills** [comment]
I found some serious bugs with certificate reloading; I have patches but didn't have chance to prepare them for upstream yet (and my other patches had no response)

Would be good to merge my bug fixes before anyone considers refactoring of the code.

**Comment by hills** [comment]
The patches in question were merged; see [this pull request](https://github.com/coturn/coturn/pull/739) so check your version is recent enough to include these patches and then you might need to file a separate issue with details if you're saying this is a problem on the latest coturn.

**Comment by cruizba** [comment]
Yes @hills, looks like it is working at the moment with latest dockerhub image.

I basically have a certbot and coturn running with these certificates. When certbot renews the certificates I am sending a SIGUSR2 signal to reload all certificates to the turnserver process. I will keep it running for a week or so to see if I still seeing crashes.

Regards.

**Comment by ziedbensalem91** [comment]
I'm using Coturn-4.5.1.1 'dan Eider' on Ubuntu 20.04.3 LTS , and try reload renewed SSL certficates without restarting the whole service to avoid downtime by using `systemctl --signal=SIGUSR2 kill coturn `, but looks like downtime happen anyway and interrupt p2p live connection

Is there another way to reload Turn SSL certificates without downtime ?

**Comment by cruizba** [comment]
Not related with your issue, but be careful with that version. It has an exploitable CVE to make a reverse proxy on other services that are running next to Coturn. I would try with the latest version available.

https://www.cvedetails.com/vulnerability-list/vendor_id-21970/product_id-70761/version_id-652851/Coturn-Project-Coturn-4.5.1.1.html

**Comment by ggarber** [comment]
@cruizba What is the status of this issue?  You mentioned you were going to keep it running for a while and see if it crashes with the latest version.   Did it crash?   Thank you very much for testing it.
